### ISSUE TITLE
Agda does not accept equational reasoning with `if_then_else` and erased invariants

### ISSUE THREAD

**Issue by HeinrichApfelmus** [agda/agda2hs#361]
## Synopsis

When trying to do the combination of

* equational reasoning
* on a definition with `if_then_else`
* on a data type which uses an additional, erased constraint

I run into problems with Agda being unable to instantiate metavariables corresponding to the `@0 {{ flg ≡ True }}` argument of the `if_then_else` construct.

In other words, Agda rejects my proof attempts when if I want to do equational reason on a definition such as

    test x1 x2 = if (x1 <= x2) then y else z

where, say, the value `y` can only be constructed by making use of the invariant `x1 <= x2 ≡ True` in the `then` branch.

## Test case

Here is a small, yet meaningful example.

First, consider a type for nonempty intervals with the constraint that the `start` of the interval always comes before the `end` of the interval:

```agda
record Interval : Set where
  field
    start : Nat
    end   : Nat
    @0 invariant : (start <= end) ≡ True
```

Here is a (stupid) test function that constructs an interval from a single value `x : Nat`.

```agda
test : Nat → Maybe Interval
test x =
  if (x <= x)
    then (λ {{@0 cond}} → Just (record
        { start = x
        ; end = x
        ; invariant = cond
        }
      ))
    else Nothing
```

I want to prove that this function always returns `Just`.

```agda
isJust : Maybe a → Bool
isJust (Just x) = True
isJust Nothing = False
```

For that, the following extended substitution principle is helpful, where `btrue` allows me to capture the condition that is valid in the branch:

```agda
prop-ifTrueEqThen
  : ∀ (b : Bool)
  → {thn : @0 {{b ≡ True}} → a}
  → {els : @0 {{b ≡ False}} → a}
  → (btrue : b ≡ True)
  → (if b then thn else els) ≡ thn {{btrue}}
prop-ifTrueEqThen .True refl = refl
```

However, trying to use this helper function results in the following problem:

```agda
prop-test
  : ∀ (x : Nat)
  → isJust (test x) ≡ True
prop-test x = cong isJust eq1
  where
    eq1 : test x ≡ Just {!   !}
    eq1 =
      ifTrueEqThen
        (x <= x)
        { ? }
        {Nothing}
        (reflexivity x)
```

    ⛔️ Cannot instantiate the metavariable _55 to solution
    record { start = x ; end = x ; invariant = cond }
    since it contains the variable cond
    which is not in scope of the metavariable
    when checking that the inferred type of an application
      (if (iOrdNat Ord.<= x) x then (λ ⦃ @0 _ ⦄ → Just ?0) else
      (λ ⦃ @0 _ ⦄ → Nothing))
      ≡ Just ?0
    matches the expected type
      test x ≡ Just ?0

## Workaround

Define a custom `if_then_else` construct where the branch is passed as a regular argument, not as an instance argument. In other words, the following is accepted by Agda:

```agda
infix -2 if'_then_else_
if'_then_else_
  : {@0 a : Set}
  → (flg : Bool)
  → (@0 (flg ≡ True) → a)
  → (@0 (flg ≡ False) → a)
  → a
if' False then x else y = y refl
if' True  then x else y = x refl

prop-if'TrueEqThen
  : ∀ (b : Bool)
  → {thn : @0 (b ≡ True) → a}
  → {els : @0 (b ≡ False) → a}
  → (btrue : b ≡ True)
  → (if' b then thn else els) ≡ (thn btrue)
prop-if'TrueEqThen .True refl = refl

test' : Nat → Maybe Interval
test' x =
  if' (x <= x)
    then (λ cond → Just (record
        { start = x
        ; end = x
        ; invariant = cond
        }
      ))
    else (λ cond → Nothing)

prop-test'
  : ∀ (x : Nat)
  → isJust (test' x) ≡ True
prop-test' x = cong isJust eq1
  where
    eq1 =
      prop-if'TrueEqThen
        (x <= x)
        (reflexivity x)
```

**Comment by jespercockx** [comment]
This is weird, when I try this example on my PC it just works (here with the holes filled):
```agda
open import Haskell.Prelude
open import Haskell.Law

record Interval : Set where
  field
    start : Nat
    end   : Nat
    @0 invariant : (start <= end) ≡ True

test : Nat → Maybe Interval
test x =
  if (x <= x)
    then (λ {{@0 cond}} → Just (record
        { start = x
        ; end = x
        ; invariant = cond
        }
      ))
    else Nothing

isJust : Maybe a → Bool
isJust (Just x) = True
isJust Nothing = False

prop-test
  : ∀ (x : Nat)
  → isJust (test x) ≡ True
prop-test x = cong isJust eq1
  where
    eq1 : test x ≡ Just (record { start = x ; end = x ; invariant = reflexivity x })
    eq1 =
      ifTrueEqThen
        (x <= x)
        { λ {{cond}} → Just (record { start = x ; end = x ; invariant = cond }) }
        {Nothing}
        (reflexivity x)
```
Which version of Agda / `agda2hs` are you using?

**Comment by HeinrichApfelmus** [comment]
I get the error on

```
% agda --version
Agda version 2.6.4.3
Built with flags (cabal -f)
 - optimise-heavily: extra optimisations
 - debug: enable debug printing ('-v' verbosity flags)
```

but I also get the error

```
⦃ cond : (iOrdNat Ord.<= x) x ≡ True ⦄ → Maybe Interval !=
Maybe Interval of type Set
when checking that the inferred type of an application
  (if (_r_93 Ord.<= x) x then
   (λ ⦃ @0 _ ⦄ ⦃ cond ⦄ →
      Just (record { start = x ; end = x ; invariant = cond }))
   else (λ ⦃ @0 _ ⦄ ⦃ cond ⦄ → Nothing))
  ≡
  (λ ⦃ cond ⦄ →
     Just (record { start = x ; end = x ; invariant = cond }))
matches the expected type
  test x ≡
  Just
  (record
   { start = x
   ; end = x
   ; invariant = IsLawfulOrd.reflexivity iLawfulOrdNat x
   })
```

on 

```
$ agda2hs --version
Agda version 2.6.4
Built with flags (cabal -f)
 - optimise-heavily: extra optimizations
 - agda2hs backend version 1.2
```

**Comment by HeinrichApfelmus** [comment]
I have set up my `agda2hs` with the following nix flake:

https://github.com/HeinrichApfelmus/agda-notes/tree/main/nix/agda-hs-tools

**Comment by omelkonian** [comment]
So this is caused by using a mismatched version of Agda for the agda2hs code, hence closing as it is not directly related to this repo.

**Comment by HeinrichApfelmus** [comment]
> So this is caused by using a mismatched version of Agda for the agda2hs code, hence closing as it is not directly related to this repo.

It's not caused by a mismatch — as indicated above, using only `agda2hs` exhibits the same problem. However, it appears that this issue has been fixed in more recent versions of `agda2hs` — but the latest *released* version is probably still affected.

**Comment by omelkonian** [comment]
@HeinrichApfelmus Yep I see what you mean, although it is slightly misleading to say that this was a bug in some previous version of `agda2hs`, since it clearly arises from the core Agda typechecker.

Just confirming that I could reproduce this with Agda-2.6.4.3, and it gets resolved in Agda-2.7.0.

**Comment by HeinrichApfelmus** [comment]
> although it is slightly misleading to say that this was a bug in some previous version of agda2hs, since it clearly arises from the core Agda typechecker.

Well, sure. But since the Agda typechecker is an inseparable part of the `agda2hs` executable version `1.2`, I am unable to act on the difference. You link against it, you hang with it. 😂
